### DISCUSSION TITLE
DynamicLinks - Handling potentially throwing `fn`

### DISCUSSION THREAD

**Discussion by marwan38** [sergiodxa/remix-utils#157] [Ideas]
`DynamicLinks` is usually placed within the `root` HTML template and there's no fail-safe in case the user input function throws for whatever reason. This might not matter to most people since the `ErrorBoundary` handles that failure, but, in our case, our `root` cannot fail because we are cheating remix and relying on the root loader data in `root` within the root ErrorBoundary. The root loader data in our case is just the `getLoadContextData` which is always available.

This is what it kinda looks like:

```tsx
// A bunch of pseudo code to demonstrate our use-case.

export const loader = ({context}) => {
  return json({context});
}

export default function Root() {
  const context = useLoaderData<typeof loader>();
  return (
    <html>
      <head>
        <DynamicLinks />
      </head>
      <body>
        <ApplicationContext context={context} />
        <Outlet />
       </body>
    </html>
  ); 
}

export default function ErrorBoundary() {
  const context = useRootLoaderData(); // I know we shouldn't be doing this, but, we haven't found a reliable work around yet.
}
```

With the following code in mind you can see that if `DynamicLinks` throws, the error boundary itself won't have what it needs which would cause it to fail as well. Wrapping that in the old school react ErrorBoundary class component was supposed to be the work-around to avoid patching and/or opening up a discussion with `remix-utils` but it doesn't work as intended for whatever reason.

So.. the question is, what your opinion @sergiodxa on adding a `try catch` around the user `fn` call in https://github.com/sergiodxa/remix-utils/blob/main/src/react/dynamic-links.tsx to ensure that it doesn't throw. I'm going to assume that this is intentional because the `ErrorBoundary` is there for this exact reason.

```tsx
export function DynamicLinks() {
    // ...
    try {
      let result = fn({
        id: match.id,
        data: match.data,
        params: match.params,
        location,
        parentsData: matches.slice(0, index).map((match) => match.data),
      });
    } catch (err) {
      console.error(err); // Probably have different/no logic based on NODE_ENV or something.
      return [];
    }
  // ...
}
```

**Comment by sciutand** [comment]
I may add that

```
export interface DynamicLinksFunction<Data extends AppData = AppData> {
  (args: HandleConventionArguments<Data>): LinkDescriptor[];
}
```
is inaccurate as `match.data` is undefined when the loader throws a (Response, Error or...) for example.

it should be

```
export interface DynamicLinksFunction<Data extends AppData = AppData> {
  (args: HandleConventionArguments<Data|undefined>): LinkDescriptor[];
}
```

or perhaps the dynamicLink function should never be called in the first place if the loader throws?

**Reply by marwan38** [reply]
I think that the latter may be out of scope for `remix-utils` since it's `remix` that handles the loaders and `useMatches` logic. The former is up to the library users.. that's why it's a generic. 

```tsx
// Consumer of `remix-utils`

const dynamicLinks: DynamiLinksFunction<Maybe<SerializeFrom<typeof loader>>> = () => { /** ... */ }
```

EDIT: Maybe i'm wrong and it's safe to always assume that `AppData` is `Maybe<AppData>` by default?

**Reply by sciutand** [reply]
Sure we can make it undefined ourselves but I'm not sure that the right approach. loader throwing (causing match.data to be undefined) is standard remix practice. The`HandleConventionArguments` type should take that in to account IMO

**Reply by sciutand** [reply]
and yes would be nice if remix would expose something like `match.didThrow` or `match.error`

**Comment by sergiodxa** [comment]
DynamicLinks are being removed in favor of using the v2 of MetaFunction.
